### PR TITLE
[refactor](dynamic table) Make segment_writer unaware of dynamic schema, and ensure parsing is exception-safe.

### PR DESCRIPTION
**Pull request by eldenmoon**
# Proposed changes

1. make ColumnObject exception safe
2. introduce FlushContext and construct schema at memtable flush stage to make segment independent from dynamic schema
3. add more test cases

Issue Number: close #xxx

## Problem summary

Describe your changes.

## Checklist(Required)

* [ ] Does it affect the original behavior
* [ ] Has unit tests been added
* [ ] Has document been added or modified
* [ ] Does it need to update dependencies
* [ ] Is this PR support rollback (If NO, please explain WHY)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by eldenmoon** [comment]
run buildall

File: be/src/olap/rowset/segment_v2/segment_writer.cpp [hunk 1]
```diff
@@ -113,37 +116,27 @@ void SegmentWriter::init_column_meta(ColumnMetaPB* meta, uint32_t column_id,
...
-// If it is a dynamic table load procedure we should handle auto generated columns.
-bool SegmentWriter::_should_create_writers_with_dynamic_block(size_t num_columns_in_block) {
-    return _tablet_schema->is_dynamic_schema() && _opts.is_direct_write &&
-           num_columns_in_block > _tablet_schema->columns().size();
+    return init(column_ids, true, flush_ctx);
 }
 
 Status SegmentWriter::init(const std::vector<uint32_t>& col_ids, bool has_key,
-                           const vectorized::Block* block) {
+                           const FlushContext* flush_ctx) {
```
Comment (yiguolei) [inline be/src/olap/rowset/segment_v2/segment_writer.cpp:132]: Add ENABLE_FACTORY_CREATOR to FlushContext to make sure smart pointer is used.

File: be/src/vec/data_types/convert_field_to_type.cpp [hunk 2]
```diff
@@ -94,8 +96,9 @@ Field convert_numeric_type_impl(const Field& from) {
     }
     return result;
 }
+
 template <typename To>
-Status convert_numric_type(const Field& from, const IDataType& type, Field* to) {
+void convert_numric_type(const Field& from, const IDataType& type, Field* to) {
```
Comment (yiguolei) [inline be/src/vec/data_types/convert_field_to_type.cpp:101]: numric? or number?

File: be/src/vec/data_types/convert_field_to_type.cpp [hunk 3]
```diff
@@ -107,18 +110,17 @@ Status convert_numric_type(const Field& from, const IDataType& type, Field* to)
     } else if (from.get_type() == Field::Types::Int128) {
         *to = convert_numeric_type_impl<Int128, To>(from);
     } else {
-        return Status::InvalidArgument(
-                fmt::format("Type mismatch in IN or VALUES section. Expected: {}. Got: {}",
-                            type.get_name(), from.get_type()));
+        throw doris::Exception(ErrorCode::INVALID_ARGUMENT,
```
Comment (yiguolei) [inline be/src/vec/data_types/convert_field_to_type.cpp:113]: are you sure segment writer or rowset writer is exception safe?

**Comment by eldenmoon** [comment]
run buildall

**Comment by eldenmoon** [comment]
run buildall

**Comment by eldenmoon** [comment]
run buildall

File: be/src/olap/rowset/segment_v2/segment_writer.cpp [hunk 4]
```diff
@@ -113,38 +116,29 @@ void SegmentWriter::init_column_meta(ColumnMetaPB* meta, uint32_t column_id,
     }
 }
 
-Status SegmentWriter::init(const vectorized::Block* block) {
+Status SegmentWriter::init(const FlushContext* flush_ctx) {
     std::vector<uint32_t> column_ids;
     int column_cnt = _tablet_schema->num_columns();
-    if (block && !_tablet_schema->is_partial_update()) {
```
[the remark this answers is not in the document]
Comment (eldenmoon) [reply be/src/olap/rowset/segment_v2/segment_writer.cpp:119]: the code is wrong, i have discussed with @yixiutt  and the code could be delete

**Review by xiaokang — APPROVED** [review]
LGTM

**Review by qidaye — APPROVED** [review]
LGTM

**Pull request merged**